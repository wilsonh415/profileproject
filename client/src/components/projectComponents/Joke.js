import React from 'react';
import { Card, Button } from 'react-bootstrap';
import laughing from '../../images/laughing.jpg';

class Joke extends React.Component {

    constructor() {
        super();
        this.state = {
            jokeSetup: null,
            jokeDelivery: null,
            joke: null,
            category: null,
            showClicked: false
        }
    }

    getJokes = async () => {
        const resp = await fetch('/api/jokes');
        const data = await resp.json();
        // need to filter out the offensive/inappropriate jokes
        const filterWords = ["sex", "penis", "FAT", "fat", "cancer", "shitty",
        "masturbating", "masturbated", "anal", "jacking", "Alzheimer's"];
        let isBad = false;
        if(data.joke !== undefined) {
            let wordCheck = filterWords.map(w => data.joke.includes(w));
            if(wordCheck.includes(true)) {
                isBad = true;
            }
        }
        if(data.setup !== undefined && data.delivery !== undefined) {
            let wordCheck = filterWords.map(w => data.setup.includes(w));
            wordCheck.push(...filterWords.map(w => data.delivery.includes(w)));
            if(wordCheck.includes(true)) {
                isBad = true;
            }
        }
        if(data.category === "Dark" || data.category === "Miscellaneous" || isBad) {
            this.getJokes();
        }
        if(data.joke !== undefined && data.category !== "Dark" 
        && data.category !== "Miscellaneous" && isBad === false) {
            this.setState({
                joke: data.joke,
                category: data.category,
                jokeDelivery: null,
                jokeSetup: null
            })
        }
        if(data.setup !== undefined && data.delivery !== undefined && isBad === false
            && data.category !== "Dark" && data.category !== "Miscellaneous") {
            this.setState({
                jokeSetup: data.setup,
                jokeDelivery: data.delivery,
                category: data.category,
                joke: null
            });
        }
    }

    componentDidMount() {
        this.getJokes();
    }

    showClicked = () => {
        this.setState({
            showClicked: true
        });
    }

    render() {
        const styles = {
            backgroundColor: "white",
            border: "1px solid",
            borderRadius: "5px",
            minWidth: "286px",
            maxWidth: "286px",
            position: "relative",
            fontFamily: "Georgia",
            marginBottom: "5vh"
        }
        return (
            <div style={styles}>
                <Card>
                    <Card.Img src={laughing} variant="top" width={286} height={180}></Card.Img>
                    <Card.Body>
                        <Card.Title><b>Random Jokes!</b></Card.Title>
                        <Card.Text>
                            Lighten up your day with a laugh!
                        </Card.Text>
                        <Card.Text style={{fontSize: "9px", marginTop: "-2vh"}}>
                            <b style={{fontSize:"10px"}}>NOTE</b>: Some jokes may be offensive and inappropriate.
                            This API was created by Sv443 (Sven Fehler)
                        </Card.Text>
                        <b><hr/></b>
                        <Card.Text style={{fontSize: "11px"}}>
                            <b>Joke Type: {this.state.category}</b>
                        </Card.Text>
                        <Card.Text>
                            {
                                (this.state.joke !== null) ? this.state.joke : null
                            }
                            <b>
                            {
                                (this.state.jokeSetup !== null) ? this.state.jokeSetup : null
                            }
                            </b>
                            <br/>
                            {
                                (this.state.jokeDelivery !== null) ? this.state.jokeDelivery : null 
                            }
                        </Card.Text>
                        <b><hr/></b>
                    </Card.Body>
                    <Button variant="primary" onClick={() => this.getJokes()}
                    style={{width: "120px", marginLeft: "29%", marginBottom: "2vh"}}>New Joke!</Button>
                </Card>
            </div>
        )
    }
}

export default Joke;