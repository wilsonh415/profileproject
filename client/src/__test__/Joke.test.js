import React from 'react';
import { shallow } from 'enzyme';
import { configure } from "enzyme";
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import Joke from '../components/projectComponents/Joke';

configure({ adapter: new Adapter() });

test('Joke component rendered properly', () => {
    const component = renderer.create(<Joke></Joke>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
