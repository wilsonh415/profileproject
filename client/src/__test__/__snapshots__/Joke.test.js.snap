// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Joke component rendered properly 1`] = `
<div
  style={
    Object {
      "backgroundColor": "white",
      "border": "1px solid",
      "borderRadius": "5px",
      "fontFamily": "Georgia",
      "marginBottom": "5vh",
      "maxWidth": "286px",
      "minWidth": "286px",
      "position": "relative",
    }
  }
>
  <div
    className="card"
  >
    <img
      className="card-img-top"
      height={180}
      src="laughing.jpg"
      width={286}
    />
    <div
      className="card-body"
    >
      <div
        className="card-title h5"
      >
        <b>
          Random Jokes!
        </b>
      </div>
      <p
        className="card-text"
      >
        Lighten up your day with a laugh!
      </p>
      <p
        className="card-text"
        style={
          Object {
            "fontSize": "9px",
            "marginTop": "-2vh",
          }
        }
      >
        <b
          style={
            Object {
              "fontSize": "10px",
            }
          }
        >
          NOTE
        </b>
        : Some jokes may be offensive and inappropriate. This API was created by Sv443 (Sven Fehler)
      </p>
      <b>
        <hr />
      </b>
      <p
        className="card-text"
        style={
          Object {
            "fontSize": "11px",
          }
        }
      >
        <b>
          Joke Type: 
        </b>
      </p>
      <p
        className="card-text"
      >
        <b />
        <br />
      </p>
      <b>
        <hr />
      </b>
    </div>
    <button
      className="btn btn-primary"
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "marginBottom": "2vh",
          "marginLeft": "29%",
          "width": "120px",
        }
      }
      type="button"
    >
      New Joke!
    </button>
  </div>
</div>
`;
